src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,253 @@
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+
+/*!
+ * Reads a value T using __ldg builtin and return a copy of it
+ */
+template<class T, typename = void>
+struct LDGLoad
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<value_type const>;
+    using reference = value_type;
+
+    CELER_CONSTEXPR_FUNCTION static reference read(pointer p)
+    {
+#if CELER_DEVICE_COMPILE
+        return __ldg(p);
+#else
+        return *p;
+#endif
+    }
+};
+
+/*!
+ * Specialization when T == OpaqueId.
+ * Wraps the underlying index in a OpaqueId when returning it.
+ */
+template<class T>
+struct LDGLoad<T, std::enable_if_t<is_opaqueid_v<T>>>
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<typename T::size_type const>;
+    using reference = value_type;
+
+    CELER_CONSTEXPR_FUNCTION static reference read(pointer p)
+    {
+#if CELER_DEVICE_COMPILE
+        return value_type{__ldg(p)};
+#else
+        return value_type{*p};
+#endif
+    }
+};
+//---------------------------------------------------------------------------//
+}  // namespace detail
+
+//---------------------------------------------------------------------------//
+/*!
+ * Iterator for read-only device data. Use __ldg intrinsic to load data in
+ * read-only cache.
+ */
+template<class T>
+class LDGIterator
+{
+    //!@{
+    //! \name Type aliases
+  private:
+    using LDGLoadPolicy = detail::LDGLoad<T>;
+
+  public:
+    using difference_type = std::ptrdiff_t;
+    using value_type = typename LDGLoadPolicy::value_type;
+    using pointer = typename LDGLoadPolicy::pointer;
+    using reference = typename LDGLoadPolicy::reference;
+    using iterator_category = std::random_access_iterator_tag;
+    //!@}
+
+  public:
+    //!@{
+    //! Construct a pointer
+    constexpr LDGIterator() noexcept = default;
+    constexpr LDGIterator(LDGIterator const&) noexcept = default;
+    CELER_CONSTEXPR_FUNCTION LDGIterator(std::nullptr_t) noexcept {}
+    CELER_CONSTEXPR_FUNCTION explicit LDGIterator(pointer ptr) noexcept
+        : ptr_{ptr}
+    {
+    }
+    //!@}
+
+    //!@{
+    //! \name RandomAccessIterator requirements
+    CELER_CONSTEXPR_FUNCTION reference operator*() const noexcept
+    {
+        return LDGLoadPolicy::read(ptr_);
+    }
+    CELER_CONSTEXPR_FUNCTION LDGIterator& operator++() noexcept
+    {
+        ++ptr_;
+        return *this;
+    }
+    CELER_CONSTEXPR_FUNCTION void swap(LDGIterator& it) noexcept
+    {
+        ::celeritas::trivial_swap(ptr_, it.ptr_);
+    }
+    CELER_CONSTEXPR_FUNCTION bool operator==(LDGIterator const& it)
+    {
+        return ptr_ == it.ptr_;
+    }
+    CELER_CONSTEXPR_FUNCTION LDGIterator operator++(int) noexcept
+    {
+        LDGIterator tmp{ptr_};
+        ++ptr_;
+        return tmp;
+    }
+    CELER_CONSTEXPR_FUNCTION pointer operator->() const noexcept
+    {
+        return ptr_;
```

Comment:
Interesting to note that if we have structs we're not going to be able to `__ldg` their data unless we return some sort of DH dereferencing wrapper to the type.

Reply:
yeah, I added this to follow the Iterator requirement. In that case, it's not useful, we could just remove it.